File: Week 6/sentimental-credit/credit.py
# Jorge Daniel Gómez Quintana "Sentimental credit"

def main():
    # Get input and validate it's a positive integer
    try:
        num = int(input("Number: "))
        if num <= 0:
            print("INVALID")
            return
    except ValueError:
        print("INVALID")
        return

    # Count digits
    length = 0
    temp = num
    while temp > 0:
        temp //= 10
        length += 1

    # Validate length
    if length not in [13, 15, 16]:
        print("INVALID")
        return

    # Get first two digits
    first_two = num
    while first_two >= 100:
        first_two //= 10
    first_digit = first_two // 10

    # Check card types
    isAmex = (length == 15) and (first_two in [34, 37])
    isMaster = (length == 16) and (51 <= first_two <= 55)
    isVisa = (first_digit == 4) and (length in [13, 16])

    # Validate card type
    if not (isAmex or isMaster or isVisa):
        print("INVALID")
        return

    # Luhn's Algorithm
    temp = num
    total = 0
    alternate = False

    while temp > 0:
        digit = temp % 10
        if alternate:
            digit *= 2
            # Add digits of products
            total += digit // 10 + digit % 10
        else:
            total += digit

        alternate = not alternate
        temp //= 10

    # Validate checksum
    if total % 10 != 0:
        print("INVALID")
        return

    # Output card type
    if isAmex:
        print("AMEX")
    elif isMaster:
        print("MASTERCARD")
    elif isVisa:
        print("VISA")


if __name__ == "__main__":
    main()

File: Week 1/credit/credit.c
// Jorge Daniel Gómez Quintana "Credit"
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long num = get_long("Number: ");
    if (num <= 0)
    {
        printf("INVALID\n");
        return 0;
    }
    // Count the number of digits
    int length = 0;
    long temp = num;
    while (temp > 0)
    {
        temp /= 10;
        length++;
    }
    // Check valid length
    if (length != 13 && length != 15 && length != 16)
    {
        printf("INVALID\n");
        return 0;
    }
    // Determine first and first two digits
    long first_two = num;
    while (first_two >= 100)
    {
        first_two /= 10;
    }
    int first_digit = first_two / 10;
    // Check card type
    int isAmex = (length == 15 && (first_two == 34 || first_two == 37));
    int isMaster = (length == 16 && (first_two >= 51 && first_two <= 55));
    int isVisa = (length == 13 || (length == 16 && first_digit == 4));

    if (!isAmex && !isMaster && !isVisa)
    {
        printf("INVALID\n");
        return 0;
    }
    // Luhn's Algorithm
    temp = num;
    int sum = 0;
    int multiply = 0;

    while (temp > 0)
    {
        int digit = temp % 10;
        if (multiply)
        {
            digit *= 2;
            sum += (digit / 10) + (digit % 10);
        }
        else
        {
            sum += digit;
        }
        multiply = !multiply;
        temp /= 10;
    }

    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    // Dteremine card type
    if (isAmex)
    {
        printf("AMEX\n");
    }
    else if (isMaster)
    {
        printf("MASTERCARD\n");
    }
    else if (isVisa)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}
